orchagent/dtelorch.cpp

```diff
@@ -0,0 +1,1384 @@
+
+    qr_attr.id = SAI_DTEL_QUEUE_REPORT_ATTR_QUEUE_ID;
+    qr_attr.value.oid = port.m_queue_ids[qreport.q_ind];
+    qreport.queue_report_attr.push_back(qr_attr);
+
+    status = sai_dtel_api->create_dtel_queue_report(&qreport.queueReportOid, 
+                gSwitchId, (uint32_t)qreport.queue_report_attr.size(), qreport.queue_report_attr.data());
+    if (status != SAI_STATUS_SUCCESS)
+    {
+        SWSS_LOG_ERROR("DTEL ERROR: Failed to enable queue report on port %s, queue %d", port.m_alias.c_str(), qreport.q_ind);
+        return status;
+    }
+
+    return status;
+}
+
+void DTelOrch::doDtelQueueReportTableTask(Consumer &consumer)
+{
+    SWSS_LOG_ENTER();
+
+    sai_status_t status;
+
+    auto it = consumer.m_toSync.begin();
+    while (it != consumer.m_toSync.end())
+    {
+        KeyOpFieldsValuesTuple t = it->second;
+        string key = kfvKey(t);
+        size_t found = key.find('|');
+        string port = key.substr(0, found);
+        string queue_id = key.substr(found + 1);
+        Port port_obj;
+        uint32_t q_ind = stoi(queue_id);
+        string op = kfvOp(t);
+
+        if (!m_portOrch->getPort(port, port_obj))
+        {
+            SWSS_LOG_ERROR("DTEL ERROR: Failed to process port for queue reporting. Port %s doesn't exist", port.c_str());
+            goto queue_report_table_continue;
+        }
+
+        if (port_obj.m_type != Port::PHY)
+        {
+            SWSS_LOG_ERROR("DTEL ERROR: Queue reporting applies only to physical ports. %s is not a physical port", port.c_str());
+   	     goto queue_report_table_continue;   
+        }
+
+        if (op == SET_COMMAND)
+        {
+            vector<sai_attribute_t> queue_report_attr;
+            sai_attribute_t qr_attr;
+            DTelQueueReportEntry *qreport;
+
+            /* If queue report is already enabled in port/queue, disable it first */
+            if (isQueueReportEnabled(port, queue_id))
+            {
+                if (!disableQueueReport(port, queue_id))
+                {
+                    goto queue_report_table_continue;
+                }
+
+                if (!removePortQueue(port, queue_id))
+                {
+                    goto queue_report_table_continue;
+                }
+            }
+
+            if (!addPortQueue(port, queue_id, &qreport))
+            {
+                goto queue_report_table_continue;
+            }
+            
+            qreport->q_ind = q_ind;
+
+            for (auto i : kfvFieldsValues(t))
+            {
+                if (fvField(i) == REPORT_TAIL_DROP)
+                {
+                    qr_attr.id = SAI_DTEL_QUEUE_REPORT_ATTR_TAIL_DROP;
+                    qr_attr.value.booldata = (fvValue(i) == ENABLED) ? true : false;
+                    qreport->queue_report_attr.push_back(qr_attr);
+                }
+                else if (fvField(i) == QUEUE_DEPTH_THRESHOLD)
+                {
+                    qr_attr.id = SAI_DTEL_QUEUE_REPORT_ATTR_DEPTH_THRESHOLD;
```

Comment:
Hi
What are the units of qdepth and latency

Reply:
qdepth is expressed in number of cells and qlatency is in nanoseconds.